dll/win32/shell32/wine/control.c

```diff
@@ -865,6 +865,105 @@ static	void	Control_DoLaunch(CPanel* panel, HWND hWnd, LPCWSTR wszCmd)
     *   "a path\foo.cpl"
     */
 {
+#ifdef __REACTOS__
+    signed sp = -1;
+    CPlApplet *applet;
+    LPCWSTR wszFirstCommaPosition = NULL, wszSecondCommaPosition = NULL;
+    LPCWSTR wszLastUnquotedSpacePosition = NULL;
+    LPWSTR buffer, buffer2;
+    BOOL bQuoted = FALSE;
+    LPCWSTR extraPmts = L"";
+    int i = 0;
+    LPWSTR ptr;
+    SIZE_T nLen = lstrlenW(wszCmd);
+
+    buffer = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(*buffer) * (nLen + 1));
+    buffer2 = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(*buffer2) * (nLen + 1));
+    if (buffer2 == NULL || buffer == NULL)
+    {
+        if (buffer != NULL)
+            HeapFree(GetProcessHeap(), 0, buffer);
+        if (buffer2 != NULL)
+            HeapFree(GetProcessHeap(), 0, buffer2);
+        return;
+    }
+
+    /* Search for unquoted commas and spaces. */
+    for (i = 0; i < nLen; i++)
+    {
+        if (wszCmd[i] == '"')
+            bQuoted = !bQuoted;
+        if (wszCmd[i] == ',' && !bQuoted)
+        {
+            if (wszFirstCommaPosition == NULL)
+                wszFirstCommaPosition = &wszCmd[i];
+            else if (wszSecondCommaPosition == NULL)
+                wszSecondCommaPosition = &wszCmd[i];
+        }
+        if (wszCmd[i] == ' ' && !bQuoted)
+        {
+            wszLastUnquotedSpacePosition = &wszCmd[i];
+        }
+    }
+
+    /* If no unquoted commas are found, parameters are either space separated, or the entire string
+     * is a path. */
+    if (wszFirstCommaPosition == NULL)
+    {
+        /* An unquoted space was found in the string. Assume the last word is the second 
+         * parameter. */
+        if (wszLastUnquotedSpacePosition != NULL)
+        {
+            int nSpaces = 0;
+
+            while (wszLastUnquotedSpacePosition[nSpaces] == L' ')
+                nSpaces++;
+
+            StringCchCopyNW(buffer, nLen, wszCmd, wszLastUnquotedSpacePosition - wszCmd);
+            lstrcpyW(buffer2, wszLastUnquotedSpacePosition + nSpaces);
+        }
+        /* No parameters were passed, the entire string is a path. */
+        else
+        {
+            lstrcpyW(buffer, wszCmd);
+        }
+    }
+    /* If an unquoted comma was found, there are at least two parts of the string:
+     * - the path 
+     * - either an index preceeded by @, or a name. 
+     * If there was a second unqoted comma, there is another part of the string:
+     * - the rest of the parameters. */
+    else
+    {
+        /* If there was no second unquoted comma in the string, the first parameter ends at the
```

Comment:
I didn't know how to call it, and maybe writing any comment in this case was a mistake. What's the correct terminology for the "@2" in "foo.cpl,@2" or "name" in "foo.cpl,name"?

Reply:
There's also a case when rundll32 can open specific property page of the CPL by its number, iirc we don't have this working in ReactOS, but it does work on Windows XP/2003, for example:
```
rundll32 shell32,Control_RunDLL desk.cpl,,0  // opens desktop page
rundll32 shell32,Control_RunDLL desk.cpl,,1  // opens screensaver page
rundll32 shell32,Control_RunDLL desk.cpl,,2  // opens visual style page
```

Reply:
The thing I called a "second parameter" is actually the thing after the first comma. I do realise now that I used a very confusing name and it will be changed once a better name is found. The stuff after the second comma is just passed to the applet procedure by the CPL anyway, and it seems to work in case of desk.cpl (maybe it didn't work before my changes, didn't check).

Reply:
Microsoft documentation of CPL messages uses the name dialog box number [(example)](https://learn.microsoft.com/en-us/windows/win32/shell/cpl-dblclk) I've changed the name of "the second parameter" to "dialog box name/number" to match that.